### DISCUSSION TITLE
`group()` aggregate function

### DISCUSSION THREAD

**Discussion by soerenwolfers** [duckdb/duckdb#14251] [Ideas]
Would be great to be able to write 
```
SELECT 
    group(), --- produces {'name': ..., 'year': ...} 
-- alternatively, groups(), -- produces two separate columns 
    sum(amount),
FROM 
    tbl
GROUP BY 
    first_name + last_name AS name,
    year
```

The query above requires two independently useful features:

1. A `group()` aggregate function (similar to but more useful than the exotic [`grouping()`](https://duckdb.org/docs/stable/sql/query_syntax/grouping_sets.html#identifying-grouping-sets-with-grouping_id) aggregate function.)
2. Being able to define aliases in the `GROUP BY` clause 

I know that `GROUP BY ALL` exists but I think it's cleaner and more maintainable to have the group definition separate from the projection. 

In fact, `GROUP BY ALL` often looks helpful but later has to be abandoned because either 

(1) the query becomes too complex for what the operator currently supports https://github.com/duckdb/duckdb/issues/18573 https://github.com/duckdb/duckdb/issues/12664 https://github.com/duckdb/duckdb/issues/14595 

(2) an edge case of `GROUP BY ALL` has a bug https://github.com/duckdb/duckdb/issues/5660 https://github.com/duckdb/duckdb/issues/18599 
https://github.com/duckdb/duckdb/issues/19239
(maybe https://github.com/duckdb/duckdb/issues/18681  )

(3) the implicitness of the `GROUP BY ALL` leads to subtle user errors in large queries https://github.com/duckdb/duckdb/issues/13470 (maybe https://github.com/duckdb/duckdb/issues/18681)

I'm undecided whether `group()` (returning a struct with all columns that are being grouped by) or `groups()` (special, expands into all columns that are being grouped by, similar to `columns(*)` expression) is better. If the former is chosen, expansion can easily be achieved with the dot-star operator. If the latter is chosen, a struct can be obtained easily by grouping by a struct expression, or via `struct_pack(*groups())`. I am leaning towards `groups()` because I think that's more frequently what users would want.

**Comment by paultiq** [comment]
I love the idea, but find 'group()' confusing, both because it creates a secondary meaning for a keyword and because it doesn't imply its behavior or purpose. 


Maybe group_columns?

**Reply by soerenwolfers** [reply]
I'm not sure whether `group()` (with parentheses! I don't like special columns) would be a problem for the parser, but I'm actually now thinking I'd like `groups()` (plural) more (so I've added it to my post).

It'd be a bit more special (one expression expanding into multiple columns) but it'd probably be more often what users want (and they could still put an explicit struct in the `group by` if they want). 

I don't like the extra verbosity in `group_columns` over `groups`.
